### ISSUE TITLE
watch fails if it does not have permission to access a dir within the 'watched' path

### ISSUE THREAD

**Issue by robjarawan** [MetPX/sarracenia#1215]
when watch (sarra) attempts to access a folder in a watch folder recursively, it fails if it does not have permission to access the directory

in this case the path to watch was: `/apps/rsshusers/home/ice/rcm/`

and the directory not accessible was `/apps/rsshusers/home/ice/rcm/.ssh/`

The log only displays a permission denied error, but i did not know that the `watch` looks into all folders recursively. 2 items i would recommend:

1- option to turn off recursive mode in watches
2- Handle better to not crashed - enhance logging - show all directories the recursive/watch is failing on in this case the .ssh/, but keep on ingesting data - Example `../home/ice/rcm/.ssh/ Permission denied.`

```2024-09-10 01:30:56,259 [DEBUG] sarracenia.flowcb.gather.file gather postpath = /apps/rsshusers/home/ice/rcm/
2024-09-10 01:30:56,259 [DEBUG] sarracenia.flowcb.gather.file watch_dir watch_dir /apps/rsshusers/home/ice/rcm/                                                               2024-09-10 01:30:56,259 [INFO] sarracenia.flowcb.gather.file watch_dir sr_watch optimal observer for platform selected (best when it works).
 2024-09-10 01:30:56,259 [INFO] sarracenia.flowcb.gather.file walk_priming sr_watch priming watch (instance=1) scheduled for: /apps/rsshusers/home/ice/rcm/
2024-09-10 01:30:56,259 [INFO] sarracenia.flowcb.gather.file watch_dir sr_watch priming walk done, but not yet active. Starting...                                            2024-09-10 01:30:56,259 [ERROR] sarracenia.flow gather flowCallback plugin <bound method File.gather of <sarracenia.flowcb.gather.file.File object at 0x7fd169b03898>> **crashed: Permission denied**
2024-09-10 01:30:56,259 [DEBUG] sarracenia.flow gather details:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sarracenia/flow/__init__.py", line 1121, in gather
    retval = p(self.o.batch-so_far)
  File "/usr/lib/python3/dist-packages/sarracenia/flowcb/gather/file.py", line 723, in gather
    messages.extend(self.watch_dir(d))
  File "/usr/lib/python3/dist-packages/sarracenia/flowcb/gather/file.py", line 662, in watch_dir
    self.observer.start()
  File "/usr/lib/python3/dist-packages/watchdog/observers/api.py", line 255, in start
    emitter.start()
  File "/usr/lib/python3/dist-packages/watchdog/utils/__init__.py", line 111, in start
    self.on_thread_start()
  File "/usr/lib/python3/dist-packages/watchdog/observers/inotify.py", line 121, in on_thread_start
    self._inotify = InotifyBuffer(path, self.watch.is_recursive)
  File "/usr/lib/python3/dist-packages/watchdog/observers/inotify_buffer.py", line 35, in __init__
    self._inotify = Inotify(path, recursive)
  File "/usr/lib/python3/dist-packages/watchdog/observers/inotify_c.py", line 187, in __init__
    self._add_dir_watch(path, recursive, event_mask)
  File "/usr/lib/python3/dist-packages/watchdog/observers/inotify_c.py", line 371, in _add_dir_watch
    self._add_watch(full_path, mask)
  File "/usr/lib/python3/dist-packages/watchdog/observers/inotify_c.py", line 385, in _add_watch
    Inotify._raise_error()
  File "/usr/lib/python3/dist-packages/watchdog/observers/inotify_c.py", line 406, in _raise_error
    raise OSError(os.strerror(err))
OSError: Permission denied```

**Comment by petersilva** [comment]
fwiw.. recursive=off was how watch originally worked, back in 2015.  It turned out "nobody" wanted that.  People kept getting disappointed when they created a directory and put files in it, and the watch would not see it.  It was basically proven that people really expect recursive to be on, even though that isn't initially obvious.  Anyways, it became not the default, and I think the option was actively removed in sr3.

I don't like the idea of adding an option just to avoid a mis-configuration... having a directory with bad permissions in the tree to be scanned is... unintuitive.  I'm ok with putting up an error message for directories we can't scan and trying to continue rather than crash... but I'd also be fine with crashing.  It's error prone, in practice to have a tree where putting files "there" won't be seen... where putting them somewhere else will be seen. Most likely the most robust solution is to create a sub-dir and scan that. e.g.
```

~/.ssh
~/transfer

```
run the watch on the transfer directory, rather than the home.

Also this issue is labelled "v3 watch" ... I think v2 is also affected in the same way, it's not version dependent.

**Comment by petersilva** [comment]
I'm not sure how *good first issue* applies here, this isn't going to be trivial to address...
```
                ow = self.observer.schedule(self.watch_handler,
                                            d,
                                            recursive=True)
  
```
means that the current code we just give it patch, and it walks the entire thing... it isn't obvious how to catcch one directory failing three levels down... or how to recover... it's just one call being done on the tree.

**Comment by reidsunderland** [comment]
I thought good first issue made sense when it was just changing the error messages to be more useful. I was hoping that it would be possible to extract more details from the Exception returned by watchdog. If that isn't possible, then I don't think we should bother trying to give more useful errors.

**Comment by reidsunderland** [comment]
Maybe it should be a separate issue, but being able to turn off recursive in a poll would be very useful. There are lots of servers that we don't control where the data we want is in one (or a few) specific directory but it has a large subdirectory structure underneath. 

We waste time (sometimes minutes, and server load/bandwidth) polling subdirectories that we don't need.

**Comment by petersilva** [comment]
yeah... ok... just wrapping the whole thing in a try/except and making a better message is not that big a deal... it would be the part where you continue to watch the rest of the tree that is the problem... the call either succeeds or fails... or it might succeed in watching the portion of the tree it scanned before it hit the problem... based on order in the directories (essentially random.)
It would not be obvious what is going on.

**Comment by petersilva** [comment]
It should probably just abort in that case.

**Comment by petersilva** [comment]
I wasn't sure if I was making stuff up... I wasn't here is the commit:

```
commit 2b862f0372d078acea97ce03425f20340ff7ef21
Author: michel grenier <michel.grenier8@canada.ca>
Date:   Wed Nov 1 21:28:38 2017 +0000

    MG sr_config take away recursive, add option statehost
       change directories according to statehost
       simplified logging

```

It doesn't say why it was removed... but if people want it back, who am I to argue... except keep the default True.

and another one...
```

commit a9269ae548be9289e1fd69be2db4bd3c27b85f1d
Author: michel grenier <michel.grenier8@canada.ca>
Date:   Wed Nov 1 21:25:29 2017 +0000

    MG sr_post force recursive True and change log settings

```

hey check this out!:

```

commit ffe40d9522dd7044388e6c90222a4eba48601443
Author: petersilva <petersilva@sourceforge.net>
Date:   Sun Sep 3 10:36:08 2017 -0400

    sr_watch:  Address huge memory usage of non-recursive inotify.
    Note to Michel: This patch switches to use built-in recursion of inotify.
    This should reduce the memory usage a lot, but now it is more brittle to failure.

```

*more brittle* ... haha!

**Comment by petersilva** [comment]
yeah so anyways story confirmed going over the git logs:
* initially nothing was recursive.
* recursive was added... everybody decided that was the "right thing" and actively removed the option.

this happenned in 2017.

**Comment by reidsunderland** [comment]
in v2 poll, the `get` option that controlled file listing on remote ftp servers might have made recursive less useful?

I'm in favour of bringing it back for polls at least. It is probably less useful in a watch.

There is one place where self.o.recursive is used in gather/file.py:

https://github.com/MetPX/sarracenia/blob/d7072428b383fa113381ddc8c65e95e574bfc856/sarracenia/flowcb/gather/file.py#L565-L574

**Comment by petersilva** [comment]
the thing is... *get* was really weird... the idea is to supply patterns to the remote server, and get a list of files as a result that match. that type of access only exists in ftp.  You end up emulating it in the protocol driver for everything else, and it just becomes a second "accept/reject" layer... so value is really limited compared to ... just using accept/reject.

**Comment by petersilva** [comment]
anyways... *get* is a whole discussion... but probably a different one.

**Comment by petersilva** [comment]
internal testing:
* the original problem is reported from using a really old version of watchdog (0.8) included with ubuntu 18.04.
* the problem is not reproducible on newer versions of watchdog installed elsewhere... which is kind of worrying... new versions just seem to silently ignore permission issues.

**Comment by petersilva** [comment]
I removed v3 from the title... because I think it's all versions that are affected.

**Comment by petersilva** [comment]
Should the fix here be to update watchdog?

**Comment by reidsunderland** [comment]
I'm not really sure what to do. Updating watchdog silently ignores the problem, which is better than not working at all, assuming you don't care about the data in the directory that the user doesn't have permission to read. 

If you did want to watch that data, then the silent failure is bad.

But it's probably a waste of time adding error handling to our code that is only needed for a really old watchdog version?

**Comment by petersilva** [comment]
probably the right answer is to contribute to watchdog and find out why they silently ignore the error... maybe contribute a patch there to raise an exception on permission errors (or an option to do so...) then once that is merged, circle back and have sr3 handle it nicely.

but that's a long haul.

**Comment by petersilva** [comment]
Also *get* ... ftp servers do globbing, so you just pass the patterns on, and you get rudimentary server side filtering.  However, that never worked on sftp because sftp doesn't do globbing. It's just ls... so you had to cd, list the whole directory, and then the sftp client did the globbing client side... which is odd because we already have accept/reject... so had two different  client-side pattern matches... which seems odd/redundant.

**Comment by petersilva** [comment]
That's why *get* is gone in sr3.

**Comment by petersilva** [comment]
also... globbing can be different on different servers... so the globbing patterns had to be tuned depending on the remote server grammar.  I don't remember the specifics, but I remember running into that.

**Comment by petersilva** [comment]
fwiw... this is sort of a duplicate of #289

**Comment by petersilva** [comment]
I tagged this a wontfix... in that we won't do anything in sarracenia to fix it... the two solutions are in watchdog.
The behaviour is understood, and there isn't much Sarracenia can do about it.
